x/wasm/types/errors.go

```diff
@@ -10,6 +10,10 @@ import (
 var (
 	DefaultCodespace = ModuleName
 
+	// ContractErrorCodes are the error codes for errors returned by the contract
+	// Since contract execution is deterministic, the errors are also deterministic
+	ContractErrorCodes = []uint32{InstantiateErrorCode, ExecuteErrorCode, QueryErrorCode, MigrateErrorCode}
```

Comment:
I think with this approach the chances are high that one of
- `ErrInstantiateFailed`
- `ErrExecuteFailed`
- `ErrQueryFailed`
- `ErrMigrationFailed`

is used in more than one place. E.g. `ErrExecuteFailed` is used many times throughout the codebase for vaious different things. `ErrMigrationFailed` is used in two places as well.

I think the safer approach is to explicitly mark errors as "Contract error" (as in the error case of

```go
// ContractResult is the raw response from the instantiate/execute/migrate calls.
// This is mirrors Rust's ContractResult<Response>.
type ContractResult struct {
	Ok  *Response `json:"ok,omitempty"`
	Err string    `json:"error,omitempty"`
}
```

in the places where we know for sure they are coming from the contract. Everything else by default remains redacted.

This would also play nicely with https://github.com/CosmWasm/wasmd/issues/1813 where we get migrations that don't even call into the contract, such that the idea of a "migration error" can be different things on two different levels.

Reply:
The term "contract error" is also used here: https://github.com/CosmWasm/cosmwasm/pull/1646

Reply:
That makes sense. So, we would add a new `ContractError` type and check against that? I was trying to do that at first, but I was not sure how to add the error codes and codespace with that.
After looking at it again, we could handle this in a similar way as the `WasmVMFlavouredError`.

Reply:
> So, we would add a new ContractError type and check against that?

That was my first idea. However, I think I missed that the error must work well in the two scenarios:
1. created in a submessage and handled in reply
2. created in a transaction requiring code and codespace

> After looking at it again, we could handle this in a similar way as the WasmVMFlavouredError.

Interesting, I did not look into that before. However, I wonder if there is a way to attach some sort of "is deterministic" property to the normal Cosmos SDK error type that we can inspect in  `redactError` to keep it simpler.

Reply:
I think that should be possible. I'll experiment a bit.

Reply:
I added a proposal in 4e7e53422c1b49fcaf482c61524cba0d88abe625. It's a wrapper type around any error that marks it as deterministic for `redactError`. I only marked the errors coming from the contract as that for now (in theory most errors produced by wasmd should be deterministic as long as we don't change the message between versions.

Reply:
Love it

> I only marked the errors coming from the contract as that for now (in theory most errors produced by wasmd should be deterministic as long as we don't change the message between versions).

We can add more and more errors over time which makes this solution extra powerful and elegant. Changing error text is usually not consensus breaking and happens in Cosmos SDK patch releases. But we can highlight that some error texts must be stable to get the extra benefit of having them non-redacted in the replies.